src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/MethodBuilderImpl.cs

```diff
@@ -49,11 +50,44 @@ internal sealed class MethodBuilderImpl : MethodBuilder
             _methodImplFlags = MethodImplAttributes.IL;
         }
 
-        internal BlobBuilder GetMethodSignatureBlob() =>
-            MetadataSignatureHelper.MethodSignatureEncoder(_module, _parameterTypes, ReturnType, !IsStatic);
+        internal BlobBuilder GetMethodSignatureBlob() => MetadataSignatureHelper.MethodSignatureEncoder(_module,
+            _parameterTypes, ReturnType, GetSignatureConvention(_callingConventions), GetGenericArguments().Length, !IsStatic);
+
+        internal static SignatureCallingConvention GetSignatureConvention(CallingConventions callingConventions)
+        {
+            // TODO: find out and handle other SignatureCallingConvention scenarios
+            SignatureCallingConvention convention = SignatureCallingConvention.Default;
+            if ((callingConventions & CallingConventions.HasThis) != 0 ||
+                (callingConventions & CallingConventions.ExplicitThis) != 0)
+            {
+                convention |= SignatureCallingConvention.ThisCall;
+            }
 
+            if ((callingConventions & CallingConventions.VarArgs) != 0)
+            {
+                convention |= SignatureCallingConvention.VarArgs;
+            }
+
+            return convention;
+        }
         protected override bool InitLocalsCore { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
-        protected override GenericTypeParameterBuilder[] DefineGenericParametersCore(params string[] names) => throw new NotImplementedException();
+        protected override GenericTypeParameterBuilder[] DefineGenericParametersCore(params string[] names)
+        {
+            if (_typeParameters != null)
+                throw new InvalidOperationException(SR.InvalidOperation_GenericParametersAlreadySet);
+
+            _typeParameters = new GenericTypeParameterBuilderImpl[names.Length];
+            for (int i = 0; i < names.Length; i++)
+            {
+                string name = names[i];
+                ArgumentNullException.ThrowIfNull(names, nameof(names));
+                _typeParameters[i] = new GenericTypeParameterBuilderImpl(name, i, this);
+            }
+
+            _isGenericMethod = true;
+            return _typeParameters;
```

Comment:
```suggestion
            var typeParameters = new GenericTypeParameterBuilderImpl[names.Length];
            for (int i = 0; i < names.Length; i++)
            {
                string name = names[i];
                ArgumentNullException.ThrowIfNull(names, nameof(names));
                typeParameters[i] = new GenericTypeParameterBuilderImpl(name, i, this);
            }
            return (_typeParameters = typeParameters);
```
Nit: It is a bit more efficient to keep the array in a local variable while populating it.

Same method on TypeBuilder can use the same change.